File: src/main/java/uk/gov/justice/hmpps/probationteams/controllers/WebMvcConfiguration.java
package uk.gov.justice.hmpps.probationteams.controllers;

import org.springframework.context.annotation.Configuration;
import org.springframework.data.web.PageableHandlerMethodArgumentResolver;
import org.springframework.data.web.SortHandlerMethodArgumentResolver;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.http.converter.StringHttpMessageConverter;
import org.springframework.web.method.support.HandlerMethodArgumentResolver;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurationSupport;

import java.util.List;

@Configuration
public class WebMvcConfiguration extends WebMvcConfigurationSupport {
    /**
     * Unregister the default {@link StringHttpMessageConverter} as we want Strings
     * to be handled by the JSON converter.
     *
     * @param converters List of already configured converters
     * @see WebMvcConfigurationSupport#addDefaultHttpMessageConverters(List)
     */
    @Override
    protected void extendMessageConverters(List<HttpMessageConverter<?>> converters) {
        converters.stream()
                .filter(c -> c instanceof StringHttpMessageConverter)
                .findFirst()
                .ifPresent(converters::remove);
    }

    /**
     * But now Spring doesn't use the method argument resolvers for Pageable and Sort.
     * Add custom PageableHandlerMethodArgumentResolver and SortHandlerMethodArgumentResolver
     * to the ones registered by default.
     * @param argumentResolvers add resolvers to this list.
     */
    @Override
    public void addArgumentResolvers(List<HandlerMethodArgumentResolver> argumentResolvers) {
        argumentResolvers.add( new PageableHandlerMethodArgumentResolver());
        argumentResolvers.add(new SortHandlerMethodArgumentResolver());
    }

}

File: src/main/java/uk/gov/justice/hmpps/probationteams/controllers/WebMvcConfiguration.kt
package uk.gov.justice.hmpps.probationteams.controllers

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.converter.HttpMessageConverter
import org.springframework.http.converter.StringHttpMessageConverter
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer

@Configuration
class WebMvcConfiguration {
    /**
     * Remove any StringHttpMessageConverters from the set configured by Spring Boot.  Now @RestControllers will correctly encode a response of type String as JSON string
     * @return The default List of HttpMessageConverter minus any StringHttpMessageConverter instances.
     */
    @Bean
    fun messageConverterReconfigurer(): WebMvcConfigurer = object : WebMvcConfigurer {
        override fun extendMessageConverters(converters: MutableList<HttpMessageConverter<*>>) {
            val filtered = converters.filter { it !is StringHttpMessageConverter }
            converters.clear()
            converters.addAll(filtered)
        }
    }
}
